h2/src/main/org/h2/mvstore/MVStore.java

```diff
@@ -1764,61 +1765,58 @@ private void compactMoveChunks(Iterable<Chunk> move) {
+            // do not overlap with space just released by chunks moved before them,
+            // hence the need to reserve this area [sourceAreaBegin, originalFileSize)
             for (Chunk chunk : move) {
-                moveChunk(chunk, sourceAreaLow, originalFileSize);
-                movedToEOF = movedToEOF || chunk.block >= originalBlokCount;
+                moveChunk(chunk, sourceAreaBegin, originalFileSize);
             }
+            // update the metadata (hopefully within the file)
+            store(sourceAreaBegin, originalFileSize);
+            sync();
 
+            Chunk chunkToMove = lastChunk;
             long biggestFileSize = getFileLengthInUse();
-            Chunk chunkToMove = null;
-            if (movedToEOF) {
-                // update the metadata (store at the end of the file)
-                reuseSpace = false;
-                store(0, biggestFileSize);
-                sync();
 
-                chunkToMove = lastChunk;
-                biggestFileSize = getFileLengthInUse();
-
-                // now re-use the empty space
-                reuseSpace = true;
-                for (Chunk c : move) {
-                    if (c.block >= originalBlokCount) {
-                        moveChunk(c, 0, 0);
-                    }
+            boolean movedToEOF = false;
+            // move all chunks, which previously did not fit before reserved area
+            // now we can re-use previously reserved area [sourceAreaBegin, originalFileSize),
+            // but need to reserve [originalFileSize, biggestFileSize)
+            for (Chunk c : move) {
+                if (c.block >= originalBlokCount) {
+                    moveChunk(c, originalFileSize, biggestFileSize);
+                    assert c.block < originalBlokCount;
+                    movedToEOF = true;
                 }
             }
-
-            // update the metadata (hopefully within the file)
-            store(movedToEOF ? originalFileSize : sourceAreaLow, biggestFileSize);
-            sync();
-
-            // if all chunks where moved straight to their target locations,
-            // but that last chunk with updated metadata did not fit,
-            // then our hope was unfounded and now we need to move it again
-            assert lastChunk != null;
-            if (!movedToEOF && lastChunk.block * BLOCK_SIZE >= biggestFileSize) {
-                chunkToMove = lastChunk;
-            }
-
-            if (chunkToMove != null &&
-                    fileStore.predictAllocation(chunkToMove.len) + chunkToMove.len < chunkToMove.block &&
-                    moveChunk(chunkToMove, 0, 0)) {
+            if ((movedToEOF || chunkToMove.block >= originalBlokCount) &&
+                    moveChunk(chunkToMove, originalFileSize, biggestFileSize)) {
+                // store a new chunk with updated metadata (hopefully within a file)
                 store(originalFileSize, biggestFileSize);
```

Comment:
Here new chunk created by store() may overloap **the old position of chunkToMove** if the chunkToMove was stored into BOF!

Reply:
I do not think, it's possible. Original location of **chunkToMove** is either less than **sourceAreaBegin** (and condition on line 1795 will prevent us from getting into this code block) or it's between **originalFileSize** and **biggestFileSize**, and we reserving that area in call to store().

Reply:
If the original location of chunkToMove is  less than sourceAreaBegin, and movedToEOF in line 1795 is true, then it's getting into this code block and bad.

Reply:
But If the original location of chunkToMove is less than sourceAreaBegin, then it will not be moved at all due to condition within moveChunkInside (1875 in newest source)